### ISSUE TITLE
Can I listen CDP events in thirtyfour?

### ISSUE THREAD

**Issue by DCjanus** [stevepryde/thirtyfour#55]
We've able to execute cdp command in `thirtyfour`, but I've not found a way to listen CDP event. 

Sometimes, I hope to use [`fetch.enable`](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#method-enable) to catch all network request in a page. When page send a network request, a `requestPaused` event is issued, and I should capture this event and send `fulfillRequest` or `failRequest`.

Like this: https://stackoverflow.com/questions/59420493/how-to-use-chrome-devtools-protocol-in-selenium-using-python-for-capturing-htt

**Comment by stevepryde** [comment]
I don't think this is possible in `thirtyfour` currently. Would you be willing to contribute this feature?

**Comment by DCjanus** [comment]
Sounds good to me.

**Comment by DCjanus** [comment]
I am not very familiar with chrome driver special protocol, like `/session/{session_id}/goog/cdp/execute`, is there any place that I can find it,

**Comment by stevepryde** [comment]
Have a look here: https://chromedevtools.github.io/devtools-protocol/

**Comment by stevepryde** [comment]
You can also check out other selenium libraries such as the python one to see how they do it

**Comment by DCjanus** [comment]
It looks like that `selenium` open a websocket connection to accept `CDP` event, which means we need to introduce a websocket library to `thirtyfour`.

Which websocket library would you like to use?

I'm not familiar with this domain, my friends told me [tokio-tungstenite](https://github.com/snapview/tokio-tungstenite) is the best chose, but it doesn't support `async-std`, and there is a [fork](https://github.com/sdroege/async-tungstenite) to support different async runtime.

**Comment by stevepryde** [comment]
Yeah have a look at https://github.com/sdroege/async-tungstenite which supports both runtimes via feature flags

**Comment by audioXD** [comment]
I thought that most libraries just use a proxy to intercept all those requests :p

**Comment by stevepryde** [comment]
You can use a proxy, however you can use CDP to do it, which I would prefer over a proxy since it's coming from the browser itself. It's more like listening to the browser's network tab

**Comment by notdanilo** [comment]
Any success on listening the events from a websocket connection? I am trying this with no success yet:

```rust
    let response = reqwest::get("http://localhost:9222/json/version")
        .await?
        .json::<HashMap<String, String>>()
        .await?;
    let ws_debugger_url = response.get("webSocketDebuggerUrl").expect("Couldn't get ws debugger url.");

    let (mut socket, _) = connect_async(Url::parse(ws_debugger_url)?).await?;
    
        let dev_tools = ChromeDevTools::new(driver.session());
    let version_info = dev_tools.execute_cdp("Browser.getVersion").await?;
    println!("{:#?}", version_info);

    std::thread::sleep(Duration::from_secs_f64(3.0));
    println!("{:#?}", dev_tools.execute_cdp("HeapProfiler.enable").await?);
    println!("{:#?}", dev_tools.execute_cdp_with_params("HeapProfiler.takeHeapSnapshot", json!({"reportProgress":true,"treatGlobalObjectsAsRoots":true,"captureNumericValue":false})).await?);

    while let Some(Ok(message)) = socket.next().await {
        println!("{}", message);
    }

```

**Comment by notdanilo** [comment]
Maybe I am missing the session ID or something that binds the websocket to the client sending the requests?

**Comment by notdanilo** [comment]
I just found this link, which describes the protocol process pretty well: https://github.com/aslushnikov/getting-started-with-cdp

**Comment by audioXD** [comment]
@notdanilo I saw that the OP made a project seperate from thirtyfour [chromiumoxide](https://github.com/DCjanus/chromiumoxide) made for `CDP`

**Comment by notdanilo** [comment]
@audioXD I urgently created a CDP client myself because I couldn't get chromiumoxide working. I was simply misunderstanding how the protocol works and chromiumoxide didn't provide a high level API for what I wanted (getting a memory heap snapshot).

I am still using thirtyfour for collecting information from html although.

**Comment by stevepryde** [comment]
For anyone else looking for CDP support, start here: https://docs.rs/thirtyfour/latest/thirtyfour/extensions/cdp/struct.ChromeDevTools.html

CDP itself is quite large, so we may not have high-level wrappers around it in the short term, but you can implement it yourself without needing to modify thirtyfour.
CDP docs are here: https://chromedevtools.github.io/devtools-protocol/

**Comment by BlinkyStitt** [comment]
> Any success on listening the events from a websocket connection? I am trying this with no success yet:
> 
>     let response = reqwest::get("http://localhost:9222/json/version")
>         .await?
>         .json::<HashMap<String, String>>()
>         .await?;
>     let ws_debugger_url = response.get("webSocketDebuggerUrl").expect("Couldn't get ws debugger url.");
> 
>     let (mut socket, _) = connect_async(Url::parse(ws_debugger_url)?).await?;
>     
>         let dev_tools = ChromeDevTools::new(driver.session());
>     let version_info = dev_tools.execute_cdp("Browser.getVersion").await?;
>     println!("{:#?}", version_info);
> 
>     std::thread::sleep(Duration::from_secs_f64(3.0));
>     println!("{:#?}", dev_tools.execute_cdp("HeapProfiler.enable").await?);
>     println!("{:#?}", dev_tools.execute_cdp_with_params("HeapProfiler.takeHeapSnapshot", json!({"reportProgress":true,"treatGlobalObjectsAsRoots":true,"captureNumericValue":false})).await?);
> 
>     while let Some(Ok(message)) = socket.next().await {
>         println!("{}", message);
>     }

I tried this out and modified it to log network responses. I think your problem is that you get the websocket url from `/json/version`. But you actually need it from `/json` and then pick the right tab that is connected to the url you care about.

Also, instead of using `dev_tools.execute_cdp`, I had to send the commands to the websocket:

```
    socket
        .send(
            json!({
                "id": 1,
                "method": "Network.enable",
            })
            .to_string()
            .into(),
        )
        .await?;
```

**Comment by MisterSzled** [comment]
I can't quite make the above work. "driver.session()" doesn't exist for instance, I'm guessing it's "driver.handle"?
Would you please create a minimal example?

**Comment by stevepryde** [comment]
CDP support has been rewritten on the main branch (not yet released on crates.io). Feel free to check it out

**Comment by stevepryde** [comment]
This is now supported. CDP support was rewritten on `main` (will ship in 0.37.0): typed commands, typed events, and event subscription via WebSocket.

Behind the `cdp-events` feature:

```rust
use thirtyfour::cdp::domains::network::*;

driver.cdp().send::<NetworkEnable>(NetworkEnable::default()).await?;
let mut requests = driver.cdp().subscribe::<NetworkRequestWillBeSent>().await?;

driver.goto("https://example.com").await?;

while let Some(evt) = requests.next().await {
    println!("{} {}", evt.request.method, evt.request.url);
}
```

For `Fetch.requestPaused` / `fulfillRequest` / `failRequest` (the original use case in this issue), the same pattern applies via the `Fetch` domain.

Examples:
- Commands & events: [thirtyfour/tests/cdp_typed.rs](https://github.com/stevepryde/thirtyfour/blob/main/thirtyfour/tests/cdp_typed.rs)
- Event subscription: [thirtyfour/tests/cdp_events.rs](https://github.com/stevepryde/thirtyfour/blob/main/thirtyfour/tests/cdp_events.rs)

Closing — please open a new issue for any specific CDP domain or event that's missing or not behaving as expected.
